### ISSUE TITLE
Roadmap

### ISSUE THREAD

**Issue by Sean-Der** [Sean-Der/opus#9]

**Comment by flamehowk** [comment]
Hello Sean.
My name is Vitaly. My program needs an Opus decoder and I would like my entire program to be in pure Go. That's why I found your Pion/Opus library. However, I can’t even check its performance, since the description does not indicate anywhere with what parameters the input.opus file must be encoded so that your decoder can decode it and turn it into output.pcm.
I have tried many different compression options, but every time I get an error: unsupported frame code: 3.
For this reason, I ask you to give me the exact encoding parameters that your decoder already supports, so that I can test its functionality.
If it works, I'm willing to try adding some missing compression options that my program needs and share them with the community (you can also include them in the Pion/Opus code).
Thank you in advance.
Sincerely,
Vitaly Monastyrskiy.

**Comment by Sean-Der** [comment]
Hi @flamehowk 

Thank you so much for reaching out! I would love to see you use this library, it has a long way to go though.

Currently it only supports CELT (The low bandwidth codec in Opus). To test that you can encode it by doing `ffmpeg -i $INPUT_FILE -c:a libopus -ac 1 -b:a 10K output.ogg`

My goal is to have both encoders working in ~3 months. I would love to work together on this if you are interested!

**Comment by flamehowk** [comment]
Hello Sean.
1) I do not need an encoder, but a decoder.
2) I need a library that I can connect to my Golang project, and not an executable codec that I will need to access through the API or in the console, because I can take a ready-made codec that is written in C ++.
3) Since your library does not have any description of the code structure, while the code is very confusing, with a superficial search, I could not understand the chain of functions that provide work with opus files. And since you did not answer for a very long time, I did just that - I just connected a codec that was compiled from source codes in C. Therefore, at the moment my problem is solved, although not in the most convenient way.

Therefore, if you send a description of the structure of this library and the algorithm of its work, so that I can understand it, then perhaps I can find the time and help you finish this codec.
However, if you do not have any descriptions of the structure of this library, then I will not spend a month of my time studying someone else's code, just to understand how it works.
Also, I will not be able to help you if, in the course of our cooperation, you will respond to my requests 2 weeks after I sent them.

I hope for your understanding.
Of good.

**Comment by Sean-Der** [comment]
1.) Great!

2.) Great!

3.) Were you able to run [decode](https://github.com/pion/opus/tree/master/examples/decode) or [playback](https://github.com/pion/opus/tree/master/examples/playback)?

 For the API we provide a [Decoder](https://github.com/pion/opus/blob/master/decoder.go#L18) and you call [Decode](https://github.com/pion/opus/blob/master/decoder.go#L26). That processes the Opus bitstream, and then calls into the SILK or CELT codecs.

I will make sure to respond faster. I am also available on the [Pion Slack](https://pion.ly/slack) and you can email me sean@pion.ly anytime. 

If you are interested in contributing I am currently just comparing this decoder to ffmpegs output. You could also look at implementing the CELT decoder. It is a pretty time intensive/tedious process though :)

```
diff --git a/libavcodec/opus_silk.c b/libavcodec/opus_silk.c
index 8523b55ada..e0ec754cdf 100644
--- a/libavcodec/opus_silk.c
+++ b/libavcodec/opus_silk.c
@@ -29,6 +29,7 @@
 #include "opus.h"
 #include "opustab.h"

+FILE *fp = NULL;
 typedef struct SilkFrame {
     int coded;
     int log_gain;
@@ -728,6 +729,11 @@ static void silk_decode_frame(SilkContext *s, OpusRangeCoder *rc,
         }
     }

+    if (fp == NULL) {
+        fp = fopen("/Users/duboisea/workspace/opus/ffmpeg-raw", "w");
+    }
+    fwrite(frame->output + SILK_HISTORY, 1, 1280, fp);
+
     frame->prev_voiced = voiced;
     memmove(frame->lpc_history, frame->lpc_history + s->flength, SILK_HISTORY * sizeof(float));
     memmove(frame->output,      frame->output      + s->flength, SILK_HISTORY * sizeof(float));
```

**Comment by flamehowk** [comment]
Hello Sean.
I already wrote to you on the e-mail that you indicated in your contacts (sean@siobud.com), but you do not answer this letter.
I can't write to you on Slack because Slack won't run on my computer.
That's why I'm writing here...
Sean, as far as I understand, you are going to rewrite source code from C to Golang alone. For this reason, you do not have the algorithms for the encoder and decoder.
Also for this reason you don't have any description of the structure of your Golang files, because you are supposed to work alone and it is enough for you to keep your project in your own head.
But, Sean, this approach to the project is not productive if you involve the participation of other colleagues. As I wrote to you in a previous letter, I don't have time to understand the C source code, and then also your Golang source code.
I asked you to specify the main function of the decoder in the Golang code so that I can (at least) see the starting point from which I can unravel the intricacies of your code. But you don't want to help.
You must understand something...
I am a refugee from Ukraine, I live in a refugee camp, I get food and 60 euros a month, so I need to solve my financial problems as soon as possible. And for this you need to finish your project and release it to the market. I don’t have time to dig into someone else’s code 2 times (first in C, then in Golang).
Therefore, forgive me, but if you do not want to help me understand your code, then I cannot help you finish it.
I wish you successful completion of your work.
All the best.

**Comment by jeanlucmongrain** [comment]
Can't wait for encoder

**Comment by gabrielfu** [comment]
thanks for the amazing package. Would love to see CELT decoder released!

**Comment by Modestying** [comment]
> Can't wait for encoder

hi, have you find other ways to resolve encoder ?

**Comment by Borisss** [comment]
great lib for opus decoder, are there any plans or ETA on encoder part? we don't need it currently but would be a nice to have option.

**Comment by JoTurk** [comment]
@Borisss Pion/opus now has a "dev" ready CELT (20ms, 48hz) and SILK encoder (mono, 20ms),  the work is done by @thomas-vilte and @fallais i think they can answer you when it will be ready for production.

**Comment by thomas-vilte** [comment]
Thanks @JoTurk for the context

@Borisss From my side, I wouldn't call it production-ready just yet. Both the CELT and SILK encoders work end-to-end (mono, 20 ms), but they're still unreleased (v0.1.0 was before all of this work)

As far as I know, I'm the only one who's really been testing them so far, and I'm still finding real bugs as I go. Just this week I fixed an allocation desync at high CBR bitrates and a pitch pre-filter bug that could double the amplitude on tonal content. There's also an MDCT fix open (#171) that came out of validating the encoder

SILK is even newer, so I still want to spend more time with that before calling it ready

I don't really want to give an ETA since I'm doing this in my spare time, but I'll post an update once I'm happy with where it's at

**Comment by fallais** [comment]
@thomas-vilte : we could try to add more tests so that it demonstrate the usage maybe ? But yes it will never replace "human ears" and real testing.

**Comment by thomas-vilte** [comment]
yeah, I agree. Right now we have `examples/decode`, `decode-webm` and `playback`, but nothing for the encoder side yet. There's no `examples/encode` and no `ExampleEncoder` in the godoc either.

the CI quality harness (`opus_compare` against libopus) is nice for catching regressions, but it doesn't really help someone just wanting to encode some audio and see how it sounds

once #171 lands, I'll put together an `examples/encode`. I'm thinking something simple: file → PCM → `Encoder.Encode` → Ogg, basically the reverse of `examples/decode`. Hopefully that'll get more people trying it out instead of everything depending on my own testing
